From 5a9867af600951203f7372d39ff3b4620dc52707 Mon Sep 17 00:00:00 2001
From: Alona Solntseva <alona.solntseva@tandemg.com>
Date: Wed, 09 Feb 2022 19:25:40 +0200
Subject: [PATCH] [SW-67906] erom pcie: port api of ppboot main

Call erom pcie API's directly instead of calling ppboot main
function
Ppboot main function ported partially
It allowes to make PCIe link up and to avoid unneessary code running in
erom. Now it runs on arc0 and doesn't prevent arcboot from continuing
to erom recovery execution
The erom_pcie image size is reduced

Change-Id: If0b77cd2c57a95507d2eb9b5dc98b69206709438
Signed-off-by: Alona Solntseva <alona.solntseva@tandemg.com>
---

diff --git a/app/ext_rom/Makefile b/app/ext_rom/Makefile
index 9ba3e86..14d7816 100644
--- a/app/ext_rom/Makefile
+++ b/app/ext_rom/Makefile
@@ -24,6 +24,7 @@
 EROM_RECOVERY_LD_SCRIPT := recovery_erom.lds
 LDFLAGS := -g -nostdlib -static
 LDFLAGS += -Wl,--gc-sections
+LDFLAGS += -Wl,--just-symbols $(PREBOOT_ROOT_DIR)/../arcboot-custom/exports.elf
 LDLIBS := -lgcc
 
 EROM_APP_S := $(shell ls crt0_common_erom.S)
@@ -35,7 +36,6 @@
 	$(PREBOOT_ROOT_DIR)/lib/common/uart.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/boot_if.o \
 	$(PREBOOT_ROOT_DIR)/lib/common/lib_common.o \
-	$(PREBOOT_ROOT_DIR)/app/ppboot/ppboot.o \
 	$(HL_DEV_ARCH)/$(HL_TARGET).o \
 	$(HL_DEV_ARCH)/../common/$(HL_TARGET).o \
 	$(HL_DEV_ARCH)/../common/common.o \
diff --git a/app/ext_rom/erom_pcie.c b/app/ext_rom/erom_pcie.c
index 97f89eb..cd400fd 100644
--- a/app/ext_rom/erom_pcie.c
+++ b/app/ext_rom/erom_pcie.c
@@ -6,15 +6,85 @@
  */
 
 #include <stdint.h>
-#include <pci.h>
+#include <hl.h>
 #include <pll.h>
 #include <preboot.h>
 #include <device.h>
+#include <uart.h>
+#include <common.h>
+#include <boot_if.h>
+
+#include <psoc_global_conf_regs.h>
+#include <pcie_aux_regs.h>
 
 static void device_init(void)
 {
+	device_init_global_conf();
 	device_init_uart();
 	device_init_pci();
+}
+
+__weak void init_prstn(void) {}
+__weak void set_pci_msix_pba_offset_val(void) {}
+__weak void preboot_init_pcie(void) {}
+__weak void ppboot_log_waiting_for_pcie(void) {}
+
+void preboot_pcie_enable(void)
+{
+	u32 rst_src;
+
+	rst_src = hl_rst_src_get();
+	switch (rst_src) {
+	case HL_COLD_RST:
+	case HL_PRSTN_RST:
+		/* PCIe is reset, so must initialize PCIe */
+		preboot_init_pcie();
+		break;
+	default:
+		/*
+		 * Currently, keep using the same flow as done before,
+		 * i.e. PCIe should already be initialized by default.
+		 * Except COLD & PRSTN resets, keep PCIe intact.
+		 */
+		break;
+	}
+}
+
+static bool hl_pcie_link_up(void)
+{
+	unsigned int prstn, smlh, rdlh, rst_src;
+
+	/* check source of reset */
+	rst_src = hl_rst_src_get();
+	if ((rst_src == HL_COLD_RST) || (rst_src == HL_PRSTN_RST))
+		return false;
+
+	prstn = regmap_readl(mmPSOC_GLOBAL_CONF_PRSTN);
+	smlh = regmap_readl(mmPCIE_AUX_SMLH_LINK_UP);
+	rdlh = regmap_readl(mmPCIE_AUX_RDLH_LINK_UP);
+
+	if (prstn && smlh && rdlh)
+		return true;
+
+	return false;
+}
+
+void hl_pcie_init(void)
+{
+	if (hl_pcie_link_up()) {
+		hl_pcie_enable_access();
+		preboot_puts("PCIe link is already UP!\n");
+		return;
+	}
+
+	if (hl_skip_pcie_init()) {
+		preboot_puts("PCIe initialization is skipped!\n");
+		return;
+	}
+
+	/* let the log be device specific according to its restrictions */
+	ppboot_log_waiting_for_pcie();
+	preboot_pcie_wait_for_link();
 }
 
 void erom_main_function(void)
@@ -27,14 +97,8 @@
 	udelay_init();
 
 	hl_uart_init();
-	init_secondary();
-	hl_pcie_init();
 
-	/*
-	 * since pcie initialization is performed by secondary
-	 * ARC (ARC0), wait here till it finishes pcie link
-	 * initialization.
-	 */
-	while (!is_pcie_initialized())
-		;
+	cpu_boot_if_init();
+	hl_plat_early_init();
+	hl_pcie_init();
 }
